// @qiita/qiita-cli/dist/commands/publish.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.publish = void 0;
const arg_1 = __importDefault(require("arg"));
const node_process_1 = __importDefault(require("node:process"));
const config_1 = require("../lib/config");
const get_file_system_repo_1 = require("../lib/get-file-system-repo");
const sync_articles_from_qiita_1 = require("../lib/sync-articles-from-qiita");
const item_validator_1 = require("../lib/validators/item-validator");
const qiita_api_1 = require("../qiita-api");
const publish = async (argv) => {
    const args = (0, arg_1.default)({
        "--all": Boolean,
    }, { argv });
    const { accessToken } = await config_1.config.getCredential();
    const qiitaApi = new qiita_api_1.QiitaApi({
        token: accessToken,
    });
    const fileSystemRepo = await (0, get_file_system_repo_1.getFileSystemRepo)();
    await (0, sync_articles_from_qiita_1.syncArticlesFromQiita)({ fileSystemRepo, qiitaApi });
    let targetItems;
    if (args["--all"]) {
        targetItems = (await fileSystemRepo.loadItems()).filter((item) => {
            return item.modified || item.id === null;
        });
    }
    else {
        const items = [];
        for (const basename of args._) {
            const item = await fileSystemRepo.loadItemByBasename(basename);
            if (item === null) {
                console.error(`Error: '${basename}' is not found`);
                node_process_1.default.exit(1);
            }
            items.push(item);
        }
        targetItems = items;
    }
    // Validate
    const invalidItemMessages = targetItems.reduce((acc, item) => {
        const errors = (0, item_validator_1.validateItem)(item);
        if (errors.length > 0)
            return [...acc, { name: item.name, errors }];
        else
            return acc;
    }, []);
    if (invalidItemMessages.length > 0) {
        console.error("Validation error:");
        invalidItemMessages.forEach((msg) => {
            console.error(msg.name, msg.errors);
            targetItems = targetItems.filter((item) => item.name !== msg.name);
        });
    }
    if (targetItems.length === 0) {
        console.log("Nothing to publish");
        node_process_1.default.exit(0);
    }
    const promises = targetItems.map(async (item) => {
        let responseItem;
        if (item.id) {
            responseItem = await qiitaApi.patchItem({
                rawBody: item.rawBody,
                tags: item.tags,
                title: item.title,
                uuid: item.id,
                isPrivate: item.secret,
                organizationUrlName: item.organizationUrlName,
            });
            console.log(`Updated: ${item.name} -> ${item.id}`);
        }
        else {
            responseItem = await qiitaApi.postItem({
                rawBody: item.rawBody,
                tags: item.tags,
                title: item.title,
                isPrivate: item.secret,
                organizationUrlName: item.organizationUrlName,
            });
            fileSystemRepo.updateItemUuid(item.name, responseItem.id);
            console.log(`Posted: ${item.name} -> ${responseItem.id}`);
        }
        await fileSystemRepo.saveItem(responseItem, false, true);
    });
    await Promise.all(promises);
    console.log("Successful!");
};
exports.publish = publish;
